### ISSUE TITLE
Flow System.Text.Rune through more APIs

### ISSUE THREAD

**Issue by GrabYourPitchforks** [dotnet/runtime#27912]
`System.Text.Rune` was recently introduced (see https://github.com/dotnet/coreclr/pull/20935) to make it easier for applications to work with Unicode text. We should do a pass of our APIs which currently take or return `System.Char` instances and enlighten them to understand `System.Text.Rune`. Here's an initial proposal that can serve as a first pass.

```cs
namespace System
{
    // Adds overloads to existing char-consuming methods
    // TODO: Do we need overloads of MemoryExtensions methods? May cause overload explosion.
    // If we had such overloads, they'd basically look like the below, but with a "this ROS<char>" parameter.
    public partial sealed class String
    {
        public bool Contains(Rune value);
        public bool Contains(Rune value, StringComparison comparisonType);

        public bool EndsWith(Rune value);
        public bool EndsWith(Rune value, StringComparison comparisonType);
        public bool EndsWith(char value, StringComparison comparisonType); // missing overload

        public int IndexOf(Rune value);
        public int IndexOf(Rune value, int startIndex);
        public int IndexOf(Rune value, int startIndex, int count);
        public int IndexOf(Rune value, StringComparison comparisonType);

        // no IndexOfAny overloads for now due to complexity of implementation, can add later if there's demand
        // no Join overloads for now, can add later if there's demand

        public int LastIndexOf(Rune value);
        public int LastIndexOf(Rune value, int startIndex);
        public int LastIndexOf(Rune value, int startIndex, int count);
        public int LastIndexOf(Rune value, StringComparison comparisonType);

        public string Replace(Rune oldRune, Rune newRune);

        public string[] Split(Rune separator, int count, StringSplitOptions options = default);
        public string[] Split(Rune separator, StringSplitOptions options = default);

        public bool StartsWith(Rune value);
        public bool StartsWith(Rune value, StringComparison comparisonType);
        public bool StartsWith(char value, StringComparison comparisonType); // missing overload

        // no Trim*(Rune[]) overloads for now due to complexity of implementation, can add later if there's demand
        public string Trim(Rune trimRune);
        public string TrimEnd(Rune trimRune);
        public string TrimStart(Rune trimRune);
    }
}

namespace System.Text
{
    // New methods on newly-introduced type
    public partial struct Rune
    {
        public static bool Equals(Rune left, Rune right); // mimics (left == right), ordinal
        public static bool Equals(Rune left, Rune right, StringComparison comparisonType); // for any non-ordinal comparison
    }

    // Adds overloads to existing char-consuming methods
    // Not all methods get overloads (e.g., AppendJoin) due to perceived rare usage and complexity of implementation
    public partial sealed class StringBuilder
    {
        public StringBuilder Append(Rune value);

        public Rune GetRuneAt(int index); // new method to match String, not an overload of an existing method

        public StringBuilder Insert(int index, Rune value);

        public StringBuilder Replace(Rune oldRune, Rune newRune);
        public StringBuilder Replace(Rune oldRune, Rune newRune, int startIndex, int count);

        public bool TryGetRuneAt(int index, out Rune value); // new method, not overload of existing method
    }
}

namespace System.IO
{
    // Adds overloads to existing char-consuming methods
    public partial abstract class TextWriter
    {
        public virtual void Write(Rune value);

        public virtual Task WriteAsync(Rune value);

        public virtual void WriteLine(Rune value);

        public virtual Task WriteLineAsync(Rune value);
    }
}

namespace System.Globalization
{
    public partial class TextInfo
    {
        public Rune ToLower(Rune r);
        public Rune ToUpper(Rune r);
    }
}
```

**Comment by GrabYourPitchforks** [comment]
Missed `CompareInfo` earlier:

```cs
namespace System.Globalization
{
    public sealed partial class CompareInfo : System.Runtime.Serialization.IDeserializationCallback
    {
        /* EXISTING APIS (partial list) */
        public int Compare(string? string1, string? string2, System.Globalization.CompareOptions options) { throw null; }
        public System.Globalization.SortKey GetSortKey(string source, System.Globalization.CompareOptions options) { throw null; }
        public int IndexOf(string source, char value, System.Globalization.CompareOptions options) { throw null; }
        public int IndexOf(string source, string value, System.Globalization.CompareOptions options) { throw null; }
        public bool IsPrefix(string source, string prefix, System.Globalization.CompareOptions options) { throw null; }
        public static bool IsSortable(char ch) { throw null; }
        public static bool IsSortable(string text) { throw null; }
        public bool IsSuffix(string source, string suffix, System.Globalization.CompareOptions options) { throw null; }
        public int LastIndexOf(string source, char value, System.Globalization.CompareOptions options) { throw null; }
        public int LastIndexOf(string source, string value, System.Globalization.CompareOptions options) { throw null; }

        /* NEW PROPOSED APIS */
        public int Compare(System.ReadOnlySpan<char> string1, System.ReadOnlySpan<char> string2, System.Globalization.CompareOptions options = System.Globalization.CompareOptions.None) { throw null; }
        public int GetSortKey(System.ReadOnlySpan<char> source, System.Span<byte> sortKey, System.Globalization.CompareOptions options = System.Globalization.CompareOptions.None) { throw null; }
        public int GetSortKeyLength(System.ReadOnlySpan<char> source, System.Globalization.CompareOptions options = System.Globalization.CompareOptions.None) { throw null; }
        public int IndexOf(System.ReadOnlySpan<char> source, System.ReadOnlySpan<char> value, System.Globalization.CompareOptions options = System.Globalization.CompareOptions.None) { throw null; }
        public int IndexOf(System.ReadOnlySpan<char> source, System.Text.Rune value, System.Globalization.CompareOptions options = System.Globalization.CompareOptions.None) { throw null; }
        public bool IsPrefix(System.ReadOnlySpan<char> source, System.ReadOnlySpan<char> prefix, System.Globalization.CompareOptions options = System.Globalization.CompareOptions.None) { throw null; }
        public static bool IsSortable(System.ReadOnlySpan<char> text) { throw null; }
        public static bool IsSortable(System.Text.Rune value) { throw null; }
        public bool IsSuffix(System.ReadOnlySpan<char> source, System.ReadOnlySpan<char> suffix, System.Globalization.CompareOptions options = System.Globalization.CompareOptions.None) { throw null; }
        public int LastIndexOf(System.ReadOnlySpan<char> source, System.ReadOnlySpan<char> value, System.Globalization.CompareOptions options = System.Globalization.CompareOptions.None) { throw null; }
        public int LastIndexOf(System.ReadOnlySpan<char> source, System.Text.Rune value, System.Globalization.CompareOptions options = System.Globalization.CompareOptions.None) { throw null; }
    }
}
```

**Comment by ericstj** [comment]
@GrabYourPitchforks are you planning to get this in for 5.0.0 or should we push it out?

**Comment by GrabYourPitchforks** [comment]
Let's push it out for now. I have a branch with most of the development done, and on the off chance it makes it for 5.0 then huzzah. :)

**Comment by jlennox** [comment]
I'd like to also request:

```
    public partial sealed class String
    {
        public string(ReadOnlySpan<Rune> runes);
    }
```

**Comment by GrabYourPitchforks** [comment]
@jlennox What scenario do you have where you end up with a `ROS<Rune>`? I don't think we expose such a construct within our libraries. But if it's common to have a sequence of these then we may want to consider API support beyond "create a string".

In the meantime, here's an \[untested\] equivalent:

```cs
static string RunesToString(ReadOnlySpan<Rune> span)
{
    // First, calculate the total required string length.
    int totalCharCount = 0;
    foreach (Rune rune in span)
    {
        totalCharCount = checked(totalCharCount + rune.Utf16SequenceLength);
    }

    // Then concat them all together.
    char[] fullBuffer = new char[totalCharCount];
    Span<char> remainingBuffer = fullBuffer;
    foreach (Rune rune in span)
    {
        remainingBuffer = remainingBuffer.Slice(rune.EncodeToUtf16(remainingBuffer));
    }

    if (!remainingBuffer.IsEmpty)
    {
        throw new Exception("Somebody modified the span while we were using it!");
    }

    return new string(fullBuffer);
}
```

If you need top-tier performance and are willing to drop down to unsafe code you can pin the incoming span and call the existing `string.Create` factory method instead.

**Comment by jlennox** [comment]
@GrabYourPitchforks 

My usecase might be a bit of a stretch. I was experimenting with the Rune API by rewriting an old swear detector that supported substitutions (foo, f00, fo0, etc) to use runes.

```C#
        public static unsafe bool IsBadWord(string word)
        {
            Span<Rune> downleveledWord = stackalloc Rune[word.Length];

            var index = 0;
            foreach (var rune in word.EnumerateRunes())
            {
                var chr = Rune.ToUpperInvariant(rune);
                var sub = rune.Value < Substitutions.Length
                    ? Substitutions[rune.Value] : (Rune)0;
                downleveledWord[index] = sub.Value != 0 ? sub : chr;
                ++index;
            }

            var s = new string(downleveledWord);
            // ...
```

**Comment by GrabYourPitchforks** [comment]
@jlennox https://www.unicode.org/reports/tr39/#Confusable_Detection might be of interest to you. It provides a mechanism to detect "confusable" characters, which are characters that look like one another. The machine-readable file that provides this data is https://www.unicode.org/Public/security/latest/confusables.txt.

As an example, check out https://util.unicode.org/UnicodeJsps/confusables.jsp?a=oO0&r=None, which provides a list of all characters that look like o, O, or 0. I've included a partial screenshot below.

![image](https://user-images.githubusercontent.com/1746272/98583639-8a344280-2279-11eb-9b5e-28b46f2270c4.png)

It's not a complete solution, but it's a start if you're curious about text processing as a generalized problem space!

**Comment by jlennox** [comment]
@GrabYourPitchforks This is really cool. Thanks!

This was actually just a coding challenge that's intended to be ASCII only :) so my moving it over to runes was really just for fun of learning the API.

**Comment by stephentoub** [comment]
@GrabYourPitchforks, @CyrusNajmabadi just asked me about StringBuilder.Append(Rune).  Should this issue be for .NET 7?

**Comment by tannergooding** [comment]
@GrabYourPitchforks is this something we can drive to completion for .NET 8? It's been approved for a while now.

**Comment by Joy-less** [comment]
This is long overdue. These overloads are necessary for Rune (and so non-BMP characters) to become a true replacement for Char. Rune overloads should be accepted anywhere Span<char> is accepted, since Rune is essentially stack-allocated char[2].

**Comment by MihaZupan** [comment]
Would you be interested in contributing the implementation?

**Comment by Joy-less** [comment]
> Would you be interested in contributing the implementation?

I've implemented every method here (and a few extra overloads) as an extension method:
https://github.com/Joy-less/System.Text.Rune-APIs
A couple of them couldn't be implemented - read the comments for more information.

**Comment by Joy-less** [comment]
> > Would you be interested in contributing the implementation?
> 
> I've implemented every method here (and a few extra overloads) as an extension method: https://github.com/Joy-less/System.Text.Rune-APIs A couple of them couldn't be implemented - read the comments for more information.

After some edits the only 2 that can't be implemented without allocations via `ToString()` are `string.Replace(Rune, Rune, StringComparison)` and `string.Split(Rune, int, StringSplitOptions)` due to the required span methods not being publicly exposed.

**Comment by MihaZupan** [comment]
What I meant was if you'd be willing to contribute an implementation _to this repo_, such that it's available out of the box for everyone.
An implementation here also has access to any internal helpers.

**Comment by Joy-less** [comment]
@MihaZupan I made an implementation in #111815. It has some compiler errors I'm not sure how to fix (`API compatibility errors between 'ref/net10.0/System.Private.CoreLib.dll' (left) and 'lib/net10.0/System.Private.CoreLib.dll' (right)`) and 2 of the methods still need to be looked at but it's a start. There's not much more I can do here.

**Comment by Joy-less** [comment]
Seeing as this has been approved, and it's been (almost entirely) implemented as a draft in #111815, could this be reviewed? @GrabYourPitchforks My pull request was automatically closed for being inactive.

**Comment by Joy-less** [comment]
@MihaZupan @stephentoub

**Comment by stephentoub** [comment]
> It has some compiler errors I'm not sure how to fix (API compatibility errors between 'ref/net10.0/System.Private.CoreLib.dll' (left) and 'lib/net10.0/System.Private.CoreLib.dll' (right)) 

You need to update the ref assemblies to match the new APIs you're adding in the implementation assemblies. e.g. https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime/ref/System.Runtime.cs

> could this be reviewed?

It won't get reviewed if it's closed.

**Comment by Joy-less** [comment]
> > It has some compiler errors I'm not sure how to fix (API compatibility errors between 'ref/net10.0/System.Private.CoreLib.dll' (left) and 'lib/net10.0/System.Private.CoreLib.dll' (right)) 
> 
> You need to update the ref assemblies to match the new APIs you're adding in the implementation assemblies. e.g. https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime/ref/System.Runtime.cs
> 
> > could this be reviewed?
> 
> It won't get reviewed if it's closed.

Okay, should I make a new pull request?

**Comment by stephentoub** [comment]
Or re-open that one if you haven't deleted the branch. Up to you.

**Comment by Joy-less** [comment]
> Or re-open that one if you haven't deleted the branch. Up to you.

It's not possible for me to re-open the pull request. I'm not a collaborator.

**Comment by Joy-less** [comment]
After implementing this proposal in #117168, I would like to suggest the following additions:

```cs
namespace System
{
    public partial sealed class String
    {
        public int IndexOf(Rune value, int startIndex, StringComparison comparisonType);
        public int IndexOf(Rune value, int startIndex, int count, StringComparison comparisonType);

        public int LastIndexOf(Rune value, int startIndex, StringComparison comparisonType);
        public int LastIndexOf(Rune value, int startIndex, int count, StringComparison comparisonType);
    }
}
```

These overloads are not in this proposal, but they are trivial to implement and match the overloads available for e.g. `IndexOf(string, int, StringComparison)` and `IndexOf(string, int, int, StringComparison)`.

I would also like to suggest the following addition:

```cs
namespace System
{
    public readonly struct Char
    {
        public bool Equals(char right, StringComparison comparisonType);
    }
}
```

~~Since this proposal is already adding `public bool StartsWith(char value, StringComparison comparisonType);`, and the above method is needed to implement this method. Even though the above can be internal, I think it would be beneficial to be public.~~

There is already `Rune.Equals(Rune, StringComparison)` so adding `char.Equals(char, StringComparison)` makes sense.

In summary, I suggest the following additions to this proposal:

```cs
namespace System
{
    public partial sealed class String
    {
        public int IndexOf(Rune value, int startIndex, StringComparison comparisonType);
        public int IndexOf(Rune value, int startIndex, int count, StringComparison comparisonType);

        public int LastIndexOf(Rune value, int startIndex, StringComparison comparisonType);
        public int LastIndexOf(Rune value, int startIndex, int count, StringComparison comparisonType);
    }

    public readonly struct Char
    {
        public bool Equals(char right, StringComparison comparisonType);
    }
}
```

**Comment by Joy-less** [comment]
@terrajobst @stephentoub I'm not sure what the process is to amend a proposal, but your input/approval/rejection of the above amendments would be greatly appreciated (especially the extra `IndexOf`/`LastIndexOf` overloads).
The amendments have already been implemented as part of my pull request in #117168, but they will be removed or left internal if not accepted.

Edit: I have read the API review process and understand that these amendments will require FXDC review.

**Comment by tarekgh** [comment]
@Joy-less please open a new issue, add the addition proposal there and then we can proceed like any other new API proposal https://github.com/dotnet/runtime/blob/main/docs/project/api-review-process.md.

**Comment by Joy-less** [comment]
@tarekgh I opened an issue in #120015. I think we should wait for the issue to be reviewed before moving forward with #117168.

**Comment by tarekgh** [comment]
> I opened an issue in https://github.com/dotnet/runtime/issues/120015. I think we should wait for the issue to be reviewed before moving forward with https://github.com/dotnet/runtime/pull/117168.

Thanks for opening the issue. We shouldn't block https://github.com/dotnet/runtime/pull/117168. We can proceed by merging this PR. Then when https://github.com/dotnet/runtime/issues/120015 is design approved, we can have another PR at that time.

**Comment by tarekgh** [comment]
Done through the PR https://github.com/dotnet/runtime/pull/117168

**Comment by jkotas** [comment]
The PR was reverted

**Comment by Joy-less** [comment]
> The PR was reverted

Please see #120145. The PR should be un-reverted now that the fixes are available.

**Comment by tarekgh** [comment]
Fixed by https://github.com/dotnet/runtime/pull/120145
